lib/surface/compiler/tokenizer.ex

```diff
@@ -415,14 +432,14 @@ defmodule Surface.Compiler.Tokenizer do
     handle_maybe_tag_open_end(text, line, column, acc, state)
   end
 
-  defp handle_attr_value_unquoted(<<c::utf8, _::binary>>, line, column, _buffer, _acc, state)
+  defp handle_attr_value_unquoted(<<c::utf8, _::binary>>, line, column, _buffer, _acc, _state)
        when c in @unquoted_value_invalid_chars do
     message = """
     unexpected character `#{<<c::utf8>>}`. \
     Unquoted attribute values cannot contain `\"`, `'`, `=` nor `<`
     """
 
-    raise %ParseError{file: state.file, line: line, column: column, message: message}
+    {:error, message, line, column}
```

Comment:
I think the same idea applies to the tokenizer. I'd just use exceptions for most cases.